src/main/java/com/google/devtools/build/lib/remote/common/BulkTransferException.java

```diff
@@ -74,31 +80,30 @@ public static boolean allCausedByCacheNotFoundException(Throwable e) {
   }
 
   /**
-   * Returns a map whose keys are the textual representation of a digest, and whose values are the
-   * corresponding action inputs
-   *
-   * <p>Use {@code Function<String, ActionInput>} to avoid the heavy dependency on {@code
-   * InputMetadataProvider}, whose getInput method provides the argument to this method.
+   * Returns a {@link LostArtifacts} instance that is non-empty if and only if all suppressed
+   * exceptions are caused by cache misses.
    */
-  public ImmutableMap<String, ActionInput> getLostInputs(
-      Function<String, ActionInput> actionInputResolver) {
+  public LostArtifacts getLostArtifacts(InputMetadataProvider metadataProvider) {
     if (!allCausedByCacheNotFoundException(this)) {
-      return ImmutableMap.of();
+      return LostArtifacts.EMPTY;
     }
 
-    ImmutableMap.Builder<String, ActionInput> lostInputs = ImmutableMap.builder();
+    var byDigestBuilder = ImmutableMap.<String, ActionInput>builder();
+    var owners = ImmutableSetMultimap.<ActionInput, Artifact>builder();
     for (var suppressed : getSuppressed()) {
       CacheNotFoundException e = (CacheNotFoundException) suppressed;
       var missingDigest = e.getMissingDigest();
       var execPath = e.getExecPath();
       checkNotNull(execPath, "exec path not known for action input with digest %s", missingDigest);
-      var actionInput = actionInputResolver.apply(execPath.getPathString());
+      var actionInput = metadataProvider.getInput(execPath.getPathString());
       checkNotNull(
           actionInput, "ActionInput not found for filename %s in CacheNotFoundException", execPath);
-
-      lostInputs.put(DigestUtil.toString(missingDigest), actionInput);
+      byDigestBuilder.put(DigestUtil.toString(missingDigest), actionInput);
     }
-    return lostInputs.buildKeepingLast();
+    var byDigest = byDigestBuilder.build();
+    return new LostArtifacts(
+        byDigest,
+        ActionRewindStrategy.calculateLostInputOwners(byDigest.values(), metadataProvider));
```

Comment:
Instead of making this visible, we can make `LostArtifacts` take an `Optional<LostInputOwners>`, and pass that into `ActionRewindStrategy`. Then the lost inputs and outputs cases would become more similar.